plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs

```diff
@@ -0,0 +1,235 @@
+    deriving stock (Show)
+
+-- See Note [Global local functions].
+-- This is one way to define the worker. There are many more, see
+-- https://github.com/input-output-hk/plutus/pull/5239#discussion_r1151197471
+-- We chose this one, because it's the simplest (no CPS shenanigans) among the safest (retrieving
+-- the next element takes O(1) time in the worst case).
+--
+-- The algorithm is a variation of the defunctionalization technique (see this post in particular:
+-- https://www.joachim-breitner.de/blog/778-Don%e2%80%99t_think,_just_defunctionalize), except we
+-- don't want a tail-recursive loop and instead emit costs lazily to the outside (as it's the whole
+-- point of the lazy costing approach)
+flattenCostRoseGo :: CostRose -> [CostRose] -> CostStream
+flattenCostRoseGo (CostRose cost1 forest1) forest2 =
+    case forest1 of
+        -- The current subtree doesn't have its own subtrees.
+        [] -> case forest2 of
+            -- No more elements in the entire tree, emit the last cost.
+            []                -> CostLast cost1
+            -- There's at least one unhandled subtree encountered earlier, emit the current cost
+            -- and collapse the unhandled subtree.
+            rose2' : forest2' -> CostCons cost1 $ flattenCostRoseGo rose2' forest2'
+        -- The current subtree has at least one its own subtree.
+        rose1' : forest1' ->
+            -- Emit the current cost and handle the list of subtrees of the current subtree.
+            CostCons cost1 $ case forest1' of
+                -- Same as the case below, except this one avoids creating a chain of
+                -- @[] ++ ([] ++ ([] ++ <...>))@, which would make retrieving the next element an
+                -- O(depth_of_the_tree) operation in the worst case.
+                [] -> flattenCostRoseGo rose1' forest2
+                -- Add the remaining subtrees of the current subtree (@forest1'@) to the stack of
+                -- all other subtrees (@forest2@) and handle the new current subtree (@rose1'@).
+                _  -> flattenCostRoseGo rose1' $ forest1' ++ forest2
+
+-- | Collapse a 'CostRose' to a lazy linear stream of costs. Retrieving the next element takes O(1)
+-- time in the worst case regardless of the recursion pattern of the given 'CostRose'.
+flattenCostRose :: CostRose -> CostStream
+flattenCostRose (CostRose cost [])              = CostLast cost
+flattenCostRose (CostRose cost (rose : forest)) = CostCons cost $ flattenCostRoseGo rose forest
+{-# INLINE flattenCostRose #-}
+
+class ExMemoryUsage a where
+    -- Inlining the implementations of this method gave us a 1-2% speedup.
+    memoryUsage :: a -> CostRose
+
+instance (ExMemoryUsage a, ExMemoryUsage b) => ExMemoryUsage (a, b) where
+    memoryUsage (a, b) = CostRose 1 [memoryUsage a, memoryUsage b]
+    {-# INLINE memoryUsage #-}
+
+instance ExMemoryUsage (SomeTypeIn uni) where
+  memoryUsage _ = CostRose 1 []
```

Comment:
> Things like `CostRose 1 []` occur quite a lot.  I was going to suggest using some kind of alias for things like this (define `constantCostRose n = CostRose n []` or something), but on second thoughts that would probably just be confusing. 

I've added 

```haskell
singletonRose :: CostingInteger -> CostRose
singletonRose cost = CostRose cost []
{-# INLINE singletonRose #-}
```

and defined those instances in terms of it. I think it's better this way: it's slightly clearer and it doesn't hardcode a particular representation of `CostRose`, which is important, because there exist other sensible ones (e.g. [this](https://github.com/input-output-hk/plutus/pull/5239/commits/803c8d9fdaf66cfe70067f03feb2f59f9494bea0) commit).